plugin/loadbalance/weighted.go

```diff
@@ -0,0 +1,299 @@
+package loadbalance
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"math/rand"
+	"net"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+type (
+	// Determinstic weighted-round-robin
+	determininsticWRR struct {
+		index int
+		count uint8
+	}
+	// Randomized weighted-round-robin
+	randomizedWRR struct {
+		wsum uint
+	}
+)
+
+func (d *determininsticWRR) nextTopIP(weights []*weightItem, rn *rand.Rand) net.IP {
```

Comment:
I'm not certain this is needed. While in theory, a deterministic implementation is different than random, IMO in the context of a server that typically serves a large number of requests, it is statistically/virtually identical to random.

Reply:
I agree. I only added this for easier testing/demonstration of the behaviour (i.e. we can check the behaviour by running only a few queries instead of 100s or 1000s of queries). If you think that is not necessary then I will remove it.

Reply:
> It is used in the unit tests, too, to verify expected results.

Technically those tests are only testing the deterministic implementation.

Reply:
True. I have removed the deterministic implementation.